### PR TITLE
added tabs and searchfilter to addpanel

### LINKED ISSUE #10427: Add search filter to new panel control
The search filter for new panels was really nice... especially when there are lots of panels (custom)

I tried a quick experiment using the modal+tabs style: 
image

It is not quite right, but the things I like are:
- x in the upper right, clear way to make the panel go away
- tabs would allow space for a more complicated copy feature.  The current 'Add to panel list' thing is a little weird

See also #10299

### PR DESCRIPTION
**Pull request by Ijin08**
fixes #10427

skarmavbild 2018-02-21 kl 14 41 53

### CODE REVIEW

File: public/app/features/dashboard/dashgrid/AddPanelPanel.tsx [hunk 1]
```diff
@@ -101,19 +119,85 @@ export class AddPanelPanel extends React.Component<AddPanelPanelProps, AddPanelP
     );
   }
 
+  filterChange(evt) {
+    this.setState({ filter: evt.target.value });
+    this.setState({ panelPlugins: this.getPanelPlugins(evt.target.value) });
```
Comment (torkelo) [inline public/app/features/dashboard/dashgrid/AddPanelPanel.tsx]: you do not need to call setState multiple times, you can pass in everything in one call

File: public/app/features/dashboard/dashgrid/AddPanelPanel.tsx [hunk 2]
```diff
@@ -92,28 +111,114 @@ export class AddPanelPanel extends React.Component<AddPanelPanelProps, AddPanelP
...
+    } else if (this.state.tab === 'Copy') {
+      addClass = '';
+      copyClass = 'active active--panel';
+      console.log(this.state.copiedPanelPlugins);
+      if (this.state.copiedPanelPlugins.length > 0) {
+        panelTab = this.state.copiedPanelPlugins.map(this.renderPanelItem);
+      } else {
+        panelTab = this.noCopiedPanelPlugins();
+      }
+    }
```
Comment (torkelo) [inline public/app/features/dashboard/dashgrid/AddPanelPanel.tsx]: this needs to refactoring (use classnames util)

**Review by torkelo — DISMISSED** [review]
Change Copy tab to Paste, and copy action in panel menu to Copy.  This component has no tests. Some tests for this component would be good.

File: public/app/features/panel/panel_ctrl.ts [hunk 3]
```diff
@@ -193,7 +193,7 @@ export class PanelCtrl {
       });
 
       menu.push({
-        text: 'Add to Panel List',
+        text: 'Copy',
         click: 'ctrl.addToPanelList()',
```
Comment (marefr) [inline public/app/features/panel/panel_ctrl.ts]: Maybe rename `addToPanelList` to `copyPanel` now? 

Another nice to have would be a keyboard shortcut (maybe P+C). But this is not a blocker for now.

**Review by marefr — CHANGES_REQUESTED** [review]
This is a really good start :+1: 

Have some questions/concerns:

**Panel area doesn't span full height:**
Maybe a result of switching scrollbar component?
image

**Positioned centered?:**
Should we position panels centered or left-aligned. Looks a bit weird with center when only showing one panel.
image

**Message needs to be changed:**
Have no good idea of that the message should be besides *Panel copied*.
image

**Paste tab:**
Since we only have at most one copied panel I think that we can skip the *Panel Search Filter* here. A custom layout of the copied panel would be nice to be able to show the full panel title. If there are no copied panel available maybe we can disable the Paste tab?
image

**Resize:**
When resizing the panel I think the content should be able to make use of the extra space, i.e. show as many panels on one row as possible?
image

**Comment by torkelo** [comment]
Since we have a paste tab would be nice to allow for like 10 last copied panels (so keep search in past tab)

File: public/app/features/panel/panel_ctrl.ts [hunk 4]
```diff
@@ -260,9 +260,9 @@ export class PanelCtrl {
     });
   }
 
-  addToPanelList() {
+  copyPanel() {
     store.set(LS_PANEL_COPY_KEY, JSON.stringify(this.panel.getSaveModel()));
-    appEvents.emit('alert-success', ['Panel temporarily added to panel list']);
+    appEvents.emit('alert-success', ['Panel copied. Open new panel to paste']);
```
Comment (marefr) [inline public/app/features/panel/panel_ctrl.ts]: We need something better here I think. How about *Panel copied, open Add Panel to paste* or similar?

**Review by marefr — CHANGES_REQUESTED** [review]
Overall looks good. A few minor things left before we can merge this (see code comment as well).

When resizing browser window the Add Panel content are not resized.

File: public/app/features/panel/panel_ctrl.ts (hunk 4 above)
Comment (torkelo) [reply public/app/features/panel/panel_ctrl.ts]: I like that, sounds more clear and gives user a direct tip on where / how to paste.

**Review by marefr — APPROVED** [review]

**Pull request merged**